sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientWrapper.scala

```diff
@@ -0,0 +1,392 @@
+      newState.err = new PrintStream(outputBuffer, true, "UTF-8")
+      newState
+    } finally {
+      Thread.currentThread().setContextClassLoader(original)
+    }
+    ret
+  }
+
+  private val client = Hive.get(conf)
+
+  private def withClassLoader[A](f: => A): A = synchronized {
+    val original = Thread.currentThread().getContextClassLoader
+    Thread.currentThread().setContextClassLoader(getClass.getClassLoader)
+    Hive.set(client)
+    version match {
+      case hive.v12 =>
+        classOf[SessionState]
+          .callStatic[SessionState, SessionState]("start", state)
+      case hive.v13 =>
+        classOf[SessionState]
+          .callStatic[SessionState, SessionState]("setCurrentSessionState", state)
+    }
+    val ret = try f finally {
+      Thread.currentThread().setContextClassLoader(original)
+    }
+    ret
+  }
+
+  def currentDatabase: String = withClassLoader {
+    state.getCurrentDatabase
+  }
+
+  def createDatabase(tableName: String): Unit = withClassLoader {
+    val table = new Table("default", tableName)
+    client.createDatabase(
+      new Database("default", "", new File("").toURI.toString, new java.util.HashMap), true)
+  }
+
+  def getDatabaseOption(name: String): Option[HiveDatabase] = withClassLoader {
+    Option(client.getDatabase(name)).map { d =>
+      HiveDatabase(
+        name = d.getName,
+        location = d.getLocationUri)
+    }
+  }
+
+  def getTableOption(dbName: String, tableName: String): Option[HiveTable] = withClassLoader {
+    logDebug(s"Looking up $dbName.$tableName")
+
+    val hiveTable = Option(client.getTable(dbName, tableName, false))
+    val converted = hiveTable.map { h =>
+
+      HiveTable(
+        name = h.getTableName,
+        specifiedDatabase = Option(h.getDbName),
+        schema = h.getCols.map(f => HiveColumn(f.getName, f.getType, f.getComment)),
+        partitionColumns = h.getPartCols.map(f => HiveColumn(f.getName, f.getType, f.getComment)),
+        properties = h.getParameters.toMap,
+        serdeProperties = h.getTTable.getSd.getSerdeInfo.getParameters.toMap,
+        tableType = ManagedTable, // TODO
+        location = version match {
+          case hive.v12 => Option(h.call[URI]("getDataLocation")).map(_.toString)
+          case hive.v13 => Option(h.call[Path]("getDataLocation")).map(_.toString)
+        },
+        inputFormat = Option(h.getInputFormatClass).map(_.getName),
+        outputFormat = Option(h.getOutputFormatClass).map(_.getName),
+        serde = Option(h.getSerializationLib)).withClient(this)
+    }
+    converted
+  }
+
+  private def toInputFormat(name: String) =
+    Class.forName(name).asInstanceOf[Class[_ <: org.apache.hadoop.mapred.InputFormat[_, _]]]
+
+  private def toOutputFormat(name: String) =
+    Class.forName(name)
+      .asInstanceOf[Class[_ <: org.apache.hadoop.hive.ql.io.HiveOutputFormat[_, _]]]
+
+  private def toQlTable(table: HiveTable): Table = {
```

Comment:
This is why I think it'd be better to prefix with hive.x.y.Table, since "Table" is so generic and could be in a lot of packages.